### PR TITLE
Add output of sections to segments mapping for ELFs 

### LINKED ISSUE #14647: rabin2 - add output of sections to segments mapping for ELFs
I suggest `-SSS` option to show the associations between ELF segments and sections, like in readelf:
```
    Elf file type is EXEC (Executable file)
    Entry point 0x1000000
    There are 6 program headers, starting at offset 64

    Program Headers:
      Type           Offset             VirtAddr           PhysAddr
		     FileSiz            MemSiz              Flags  Align
      LOAD           0x0000000000200000 0xffffffff81000000 0x0000000001000000
		     0x00000000007a3000 0x00000000007a3000  R E    200000
      LOAD           0x0000000000a00000 0xffffffff81800000 0x0000000001800000
		     0x00000000000c7b40 0x00000000000c7b40  RW     200000
      LOAD           0x0000000000c00000 0xffffffffff600000 0x00000000018c8000
		     0x0000000000000d60 0x0000000000000d60  R E    200000
      LOAD           0x0000000000e00000 0x0000000000000000 0x00000000018c9000
		     0x0000000000010f40 0x0000000000010f40  RW     200000
      LOAD           0x0000000000eda000 0xffffffff818da000 0x00000000018da000
		     0x0000000000095000 0x0000000000163000  RWE    200000
      NOTE           0x0000000000713e08 0xffffffff81513e08 0x0000000001513e08
		     0x0000000000000024 0x0000000000000024         4

     Section to Segment mapping:
      Segment Sections...
       00     .text .notes __ex_table .rodata __bug_table .pci_fixup __ksymtab 
	      __ksymtab_gpl __ksymtab_strings __init_rodata __param __modver 
       01     .data 
       02     .vsyscall_0 .vsyscall_fn .vsyscall_1 .vsyscall_2 .vsyscall_var_jiffies 
	      .vsyscall_var_vgetcpu_mode .vsyscall_var_vsyscall_gtod_data 
       03     .data..percpu 
       04     .init.text .init.data .x86_trampoline .x86_cpu_dev.init .altinstructions 
	      .altinstr_replacement .iommu_table .apicdrivers .exit.text .smp_locks 
	      .data_nosave .bss .brk 
       05     .notes 
```

### PR DESCRIPTION
**Pull request by kuqadk3**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

Merging this pull request will allow the output of sections to segments mapping for ELFs in rabin2 with "-SSS" option. In which, solve issue #14647 
...

**Test plan**

...

**Closing issues**
closes #14647 

...

### CODE REVIEW

File: libr/main/rabin2.c [hunk 1]
```diff
@@ -471,6 +471,48 @@ static int rabin_show_srcline(RBin *bin, ut64 at) {
...
 
+/* Map Sections to Segments https://github.com/radareorg/radare2/issues/14647 */
+static int rabin_map_sections_to_segments(RBin *bin){
+	RListIter *iter;
+	RBinSection *section = NULL, *segment = NULL;
+	RList *sections = r_list_new();
+	RList *segments = r_list_new();
+	RList *tmp = r_bin_get_sections (bin);
+	printf ("Section to Segment mapping:\n");
+	printf ("Segment\tSections...\n");
```
Comment (radare) [inline libr/main/rabin2.c]: dont use \t

File: libr/main/rabin2.c [hunk 2]
```diff
@@ -471,6 +471,48 @@ static int rabin_show_srcline(RBin *bin, ut64 at) {
...
+		} else {
+			segment = r_list_pop_head (segments);
+			segment_itv = (RInterval){segment->vaddr, segment->size};
+			printf ("\n");
+			printf ("%s\t", segment->name);
+		}
+	}
+	printf ("\n");
+	/* print out left-over segments that contains no sections */
+	while (r_list_length(segments) > 0) {
```
Comment (radare) [inline libr/main/rabin2.c]: missing space

File: libr/main/rabin2.c [hunk 3]
```diff
@@ -471,6 +471,48 @@ static int rabin_show_srcline(RBin *bin, ut64 at) {
 	return false;
 }
 
+/* Map Sections to Segments https://github.com/radareorg/radare2/issues/14647 */
+static int rabin_map_sections_to_segments(RBin *bin){
+	RListIter *iter;
+	RBinSection *section = NULL, *segment = NULL;
+	RList *sections = r_list_new();
```
Comment (radare) [inline libr/main/rabin2.c]: mssing space before (

File: libr/main/rabin2.c [hunk 4]
```diff
@@ -471,6 +471,48 @@ static int rabin_show_srcline(RBin *bin, ut64 at) {
 	return false;
 }
 
+/* Map Sections to Segments https://github.com/radareorg/radare2/issues/14647 */
+static int rabin_map_sections_to_segments(RBin *bin){
```
Comment (radare) [inline libr/main/rabin2.c]: if its internal prefx with __

File: libr/main/rabin2.c [hunk 5]
```diff
@@ -471,6 +471,48 @@ static int rabin_show_srcline(RBin *bin, ut64 at) {
...
+		if (section->is_segment) {
+			r_list_append (segments, section);
+		} else {
+			r_list_append (sections, section);
+		}
+	}
+
+	segment = r_list_pop_head (segments);
+	RInterval segment_itv = (RInterval){segment->vaddr, segment->size};
+	printf ("%s\t", segment->name);
```
Comment (radare) [inline libr/main/rabin2.c]: dont use \t

File: libr/main/rabin2.c [hunk 6]
```diff
@@ -471,6 +471,48 @@ static int rabin_show_srcline(RBin *bin, ut64 at) {
...
+	RListIter *iter;
+	RBinSection *section = NULL, *segment = NULL;
+	RList *sections = r_list_new();
+	RList *segments = r_list_new();
+	RList *tmp = r_bin_get_sections (bin);
+	printf ("Section to Segment mapping:\n");
+	printf ("Segment\tSections...\n");	
+
+	r_list_foreach(tmp, iter, section) {
+		if (section->is_segment) {
```
Comment (radare) [inline libr/main/rabin2.c]: ideally we should have separae callbacks to collect sections and segments

File: libr/main/rabin2.c (hunk 2 above)
Comment (radare) [reply libr/main/rabin2.c]: better use while (!r_list_empty(segments))

File: libr/main/rabin2.c [hunk 7]
```diff
@@ -471,6 +471,48 @@ static int rabin_show_srcline(RBin *bin, ut64 at) {
...
+			segment_itv = (RInterval){segment->vaddr, segment->size};
+			printf ("\n");
+			printf ("%s\t", segment->name);
+		}
+	}
+	printf ("\n");
+	/* print out left-over segments that contains no sections */
+	while (r_list_length(segments) > 0) {
+		segment = r_list_pop_head (segments);
+		printf ("%s\t\n", segment->name);
```
Comment (radare) [inline libr/main/rabin2.c]: use table api instead of printfs with tabs or at least %20s and such, but tabs are not reliable for consistent output across editors/visualizers/shells/fonts/.. its better not to use them

File: libr/main/rabin2.c [hunk 8]
```diff
@@ -704,8 +747,10 @@ R_API int r_main_rabin2(int argc, char **argv) {
 			if (is_active (R_BIN_REQ_SECTIONS)) {
 				action &= ~R_BIN_REQ_SECTIONS;
 				action |= R_BIN_REQ_SEGMENTS;
+				sss += 1;
```
Comment (radare) [inline libr/main/rabin2.c]: use ++

File: libr/main/rabin2.c [hunk 9]
```diff
@@ -704,8 +747,10 @@ R_API int r_main_rabin2(int argc, char **argv) {
...
 			} else {
 				set_action (R_BIN_REQ_SECTIONS);
+				sss += 1;
```
Comment (radare) [inline libr/main/rabin2.c]: same operation in if and else. just put that statement below outside the conditional blocks

File: libr/main/rabin2.c [hunk 10]
```diff
@@ -1164,6 +1209,9 @@ R_API int r_main_rabin2(int argc, char **argv) {
 				"plugins found!\n", bin->file);
 		}
 	}
+	if (sss == 3) {
```
Comment (radare) [inline libr/main/rabin2.c]: uhm, -SSS right now is the same as passing -S -SS, not sure if that will be what we want :?

File: libr/main/rabin2.c (hunk 6 above)
Comment (kuqadk3) [reply libr/main/rabin2.c]: True, I was also quite confused when I figured out r_bin_get_sections() return both sections and segments. But to change that to two separate calls, it requires lots of work and changes a lot of code that uses is_segment...and I think it worth opening a new issue about this

https://github.com/radareorg/radare2/blob/80c74d7aa72e3bb4561e0ecba1f1e6b28599db54/libr/bin/p/bin_symbols.c#L68-L84

File: libr/main/rabin2.c [hunk 11]
```diff
@@ -471,6 +471,48 @@ static int rabin_show_srcline(RBin *bin, ut64 at) {
 	return false;
 }
 
+/* Map Sections to Segments https://github.com/radareorg/radare2/issues/14647 */
+static int __rabin_map_sections_to_segments(RBin *bin) {
```
Comment (ret2libc) [inline libr/main/rabin2.c]: I believe that any function in rabin2 should just print information obtained by the libraries. By implementing all the logic in this file, you can get this functionality only through rabin2. I think a better approach is to implement a new API in RBin, something like:
```C
struct r_bin_segment_section_map {
   RBinSection *segment;
   RList *sections;
} RBinSegmentSectionMap;
RList/*RBinSegmentSectionMap*/ *r_bin_segment_sections_map(RBin *bin); 
```
and then call this function and just print the results however you want.

File: libr/main/rabin2.c (hunk 10 above)
Comment (ret2libc) [reply libr/main/rabin2.c]: Actually... this is true anyway. And also `-SS` is equal to `-S -S` right now.

**Comment by ret2libc** [comment]
Also, please add a test for this new feature.

File: libr/main/rabin2.c (hunk 10 above)
Comment (kuqadk3) [reply libr/main/rabin2.c]: Yeah, and -SSS is already equal to -S -SS even without this.

File: libr/main/rabin2.c (hunk 11 above)
Comment (ret2libc) [reply libr/main/rabin2.c]: Or, even better, we should probably use something similar to what's done with Sections, Segments, etc.

See R_BIN_REQ_SECTIONS/R_CORE_BIN_ACC_SECTIONS, R_BIN_REQ_SEGMENTS/R_CORE_BIN_ACC_SEGMENTS, etc.

Probably this functionality should also be implemented in `r_core_bin_info`.

File: libr/main/rabin2.c [hunk 12]
```diff
@@ -471,6 +471,48 @@ static int rabin_show_srcline(RBin *bin, ut64 at) {
...
+		} else {
+			r_list_append (sections, section);
+		}
+	}
+
+	segment = r_list_pop_head (segments);
+	RInterval segment_itv = (RInterval){segment->vaddr, segment->size};
+	printf ("%s\t", segment->name);
+
+	r_list_foreach (sections, iter, section) {
```
[the remark this answers is not in the document]
Comment (kuqadk3) [reply libr/main/rabin2.c]: Sorry, I thought r_bin_get_sections() will return sections in order. Then we probably need two for loops. Gonna fix it, thanks!

[the remark this answers is not in the document]
Comment (kuqadk3) [reply libr/main/rabin2.c]: Fixed

**Comment by kuqadk3** [comment]
Please review @radare @ret2libc . Thanks!

File: libr/main/rabin2.c (hunk 10 above)
Comment (kuqadk3) [reply libr/main/rabin2.c]: Thanks for the code! 🍻

File: libr/core/cbin.c [hunk 13]
```diff
@@ -2491,6 +2491,41 @@ struct io_bin_section_info_t {
 	int fd;
 };
 
+/* Map Sections to Segments https://github.com/radareorg/radare2/issues/14647 */
+static int rabin_map_sections_to_segments(RBin *bin) {
```
Comment (ret2libc) [inline libr/core/cbin.c]: rename to `bin_map_sections_to_segments`

Comment (ret2libc) [reply libr/core/cbin.c]: or something better... in any case, `rabin_` prefix is not ok.

File: libr/main/rabin2.c [hunk 14]
```diff
@@ -471,6 +471,41 @@ static int rabin_show_srcline(RBin *bin, ut64 at) {
 	return false;
 }
 
+
```
Comment (ret2libc) [inline libr/main/rabin2.c]: remove all this code.

File: libr/main/rabin2.c [hunk 15]
```diff
@@ -1151,6 +1189,7 @@ R_API int r_main_rabin2(int argc, char **argv) {
 	run_action ("versioninfo", R_BIN_REQ_VERSIONINFO, R_CORE_BIN_ACC_VERSIONINFO);
 	run_action ("sections", R_BIN_REQ_SIGNATURE, R_CORE_BIN_ACC_SIGNATURE);
 	run_action ("hashes", R_BIN_REQ_HASHES, R_CORE_BIN_ACC_HASHES);
+	run_action ("sections mapping", R_BIN_REQ_SECTIONS_MAPPING, R_CORE_BIN_ACC_SECTIONS_MAPPING);
```
Comment (ret2libc) [inline libr/main/rabin2.c]: I think it's useful to add `R_BIN_REQ_SECTIONS_MAPPING` at https://github.com/radareorg/radare2/blob/master/libr/main/rabin2.c#L652 as well. Also, please update the help message at https://github.com/radareorg/radare2/blob/master/libr/main/rabin2.c#L32

File: libr/core/cbin.c [hunk 16]
```diff
@@ -3888,6 +3923,9 @@ R_API int r_core_bin_info(RCore *core, int action, int mode, int va, RCoreBinFil
 	if ((action & R_CORE_BIN_ACC_SEGMENTS)) {
 		ret &= bin_sections (core, mode, loadaddr, va, at, name, chksum, true);
 	}
+	if ((action & R_CORE_BIN_ACC_SECTIONS_MAPPING)) {
+		ret &= rabin_map_sections_to_segments(core->bin);
```
Comment (ret2libc) [inline libr/core/cbin.c]: Are you willing to implement the json mode as well? As you can see, other actions in this function accept a `mode` argument, so they can print in different modes. There is no need, in this case, to have the other modes, but the json one is usually useful especially for r2pipe.

If not, it's ok anyway :)

File: libr/core/cbin.c [hunk 17]
```diff
@@ -2491,6 +2491,41 @@ struct io_bin_section_info_t {
...
 
+/* Map Sections to Segments https://github.com/radareorg/radare2/issues/14647 */
+static int rabin_map_sections_to_segments(RBin *bin) {
+	RListIter *iter, *iter2;
+	RBinSection *section = NULL, *segment = NULL;
+	RList *sections = r_list_new();
+	RList *segments = r_list_new();
+	RList *tmp = r_bin_get_sections (bin);
+
+	printf ("Section to Segment mapping:\n");
```
Comment (ret2libc) [inline libr/core/cbin.c]: I believe @radare suggested to use RTable api, right? I'm not sure how that would work in this case, have you tried it?

File: libr/main/rabin2.c [hunk 18]
```diff
@@ -57,6 +57,7 @@ static int rabin_show_help(int v) {
 		" -s              symbols\n"
 		" -S              sections\n"
 		" -SS             segments\n"
+		" -SSS            sections mapping to segments\n"
```
Comment (ret2libc) [inline libr/main/rabin2.c]: add `-SSS` to line 32 as well, as `-g` option is a way to say "all flags" and -SSS should be listed there as well since we have added R_BIN_REQ_SECTIONS_MAPPING to the 'g' flag.

File: libr/core/cbin.c (hunk 17 above)
Comment (kuqadk3) [reply libr/core/cbin.c]: It uses RTable API now, thanks for the suggestion.

File: libr/core/cbin.c [hunk 19]
```diff
@@ -2491,6 +2491,46 @@ struct io_bin_section_info_t {
 	int fd;
 };
 
+/* Map Sections to Segments https://github.com/radareorg/radare2/issues/14647 */
+static int bin_map_sections_to_segments(RBin *bin) {
+	RListIter *iter, *iter2;
+	RBinSection *section = NULL, *segment = NULL;
+	RList *sections = r_list_new();
```
Comment (radare) [inline libr/core/cbin.c]: missing space before (

File: libr/core/cbin.c [hunk 20]
```diff
@@ -2491,6 +2491,46 @@ struct io_bin_section_info_t {
...
+
+	printf ("Section to Segment mapping:\n");
+	RTable *table = r_table_new();
+	RTableColumnType *typeString = r_table_type ("string");
+
+	r_table_add_column (table, typeString, "Segment", 0);
+	r_table_add_column (table, typeString, "Section", 0);
+
+	r_list_foreach(tmp, iter, section) {
+		if (section->is_segment) {
```
Comment (radare) [inline libr/core/cbin.c]: Use the ternary operator:

RList *list = section->is_segment? segments: sections;
r_list_append(list, section);

File: libr/core/cbin.c [hunk 21]
```diff
@@ -2491,6 +2491,46 @@ struct io_bin_section_info_t {
...
+		r_list_foreach (sections, iter2, section) {
+			RInterval section_itv = (RInterval){section->vaddr, section->size};
+			if (r_itv_begin (section_itv) >= r_itv_begin (segment_itv) && r_itv_end (section_itv) <= r_itv_end (segment_itv)) {
+				tmp2 = r_str_appendf (tmp2, "%s ", section->name);
+			}
+		}
+		r_table_add_row(table, segment->name, tmp2, 0);
+	}
+
+	printf("%s", r_table_tostring(table));
```
Comment (radare) [inline libr/core/cbin.c]: missing space before (

Comment (radare) [inline libr/core/cbin.c]: use r_cons_printf not printf

File: libr/core/cbin.c [hunk 22]
```diff
@@ -2491,6 +2491,46 @@ struct io_bin_section_info_t {
...
+	printf ("\n");
```
Comment (radare) [inline libr/core/cbin.c]: use r_cons_newline ()

File: libr/core/cbin.c [hunk 23]
```diff
@@ -2491,6 +2491,46 @@ struct io_bin_section_info_t {
...
+	r_list_foreach (segments, iter, segment) {
+		RInterval segment_itv = (RInterval){segment->vaddr, segment->size};
+		tmp2 = r_str_new ("");
+		r_list_foreach (sections, iter2, section) {
+			RInterval section_itv = (RInterval){section->vaddr, section->size};
+			if (r_itv_begin (section_itv) >= r_itv_begin (segment_itv) && r_itv_end (section_itv) <= r_itv_end (segment_itv)) {
+				tmp2 = r_str_appendf (tmp2, "%s ", section->name);
+			}
+		}
+		r_table_add_row(table, segment->name, tmp2, 0);
```
Comment (radare) [inline libr/core/cbin.c]: missing space before (

File: libr/core/cbin.c [hunk 24]
```diff
@@ -2491,6 +2491,46 @@ struct io_bin_section_info_t {
...
+	RList *tmp = r_bin_get_sections (bin);
+
+	printf ("Section to Segment mapping:\n");
+	RTable *table = r_table_new();
+	RTableColumnType *typeString = r_table_type ("string");
+
+	r_table_add_column (table, typeString, "Segment", 0);
+	r_table_add_column (table, typeString, "Section", 0);
+
+	r_list_foreach(tmp, iter, section) {
```
Comment (radare) [inline libr/core/cbin.c]: missing space before (

File: libr/core/cbin.c [hunk 25]
```diff
@@ -2491,6 +2491,46 @@ struct io_bin_section_info_t {
...
+/* Map Sections to Segments https://github.com/radareorg/radare2/issues/14647 */
+static int bin_map_sections_to_segments(RBin *bin) {
+	RListIter *iter, *iter2;
+	RBinSection *section = NULL, *segment = NULL;
+	RList *sections = r_list_new();
+	RList *segments = r_list_new();
+	RList *tmp = r_bin_get_sections (bin);
+
+	printf ("Section to Segment mapping:\n");
+	RTable *table = r_table_new();
```
Comment (radare) [inline libr/core/cbin.c]: missing space before (

File: libr/core/cbin.c (hunk 17 above)
Comment (radare) [reply libr/core/cbin.c]: use r_cons_printf for now

**Comment by kuqadk3** [comment]
Sorry for those tons of space fixing stuff...but seems like clang-format-diff.py could not detect those mistakes or I am wrong some where?

```
tuanlinh@tuanlinh-virtual-machine:~/Desktop/radare2-gsoc/radare2$ git diff master .|./sys/clang-format-diff.py -p1                                                                                                 
tuanlinh@tuanlinh-virtual-machine:~/Desktop/radare2-gsoc/radare2$
--


```

File: libr/core/cbin.c (hunk 16 above)
Comment (kuqadk3) [reply libr/core/cbin.c]: Nice, let me try.

Comment (kuqadk3) [reply libr/core/cbin.c]: Added. Please review, thanks!

**Comment by ret2libc** [comment]
> Sorry for those tons of space fixing stuff...but seems like clang-format-diff.py could not detect those mistakes or I am wrong some where?

It could be the script does not work perfectly... As it is really hard to have a tool that formats automatically radare2 code base the way "we" want it, the tool just tries to help by detecting wrong changes instead of formatting entire files.

So, it should work just on the modified lines of the output `git diff master..mybranch`. If a line was not modified (thus it does not appear in the diff output, it won't be checked.

File: libr/core/cbin.c [hunk 26]
```diff
@@ -2491,6 +2491,52 @@ struct io_bin_section_info_t {
...
+		r_list_append (list, section);
+	}
+
+	char *tmp2 = NULL;
+	r_list_foreach (segments, iter, segment) {
+		RInterval segment_itv = (RInterval){segment->vaddr, segment->size};
+		tmp2 = r_str_new ("");
+		r_list_foreach (sections, iter2, section) {
+			RInterval section_itv = (RInterval){section->vaddr, section->size};
+			if (r_itv_begin (section_itv) >= r_itv_begin (segment_itv) && r_itv_end (section_itv) <= r_itv_end (segment_itv) && strlen (section->name) > 0) {
```
Comment (ret2libc) [inline libr/core/cbin.c]: there is no need to compute the entire length just to know if > 0.

You can do `section->name[0]` and that's faster.

File: package.json [hunk 27]
```diff
@@ -24,5 +24,8 @@
   "version": "1.6.0-git",
   "bugs": {
     "url": "https://github.com/radareorg/radare2/issues"
+  },
+  "dependencies": {
+    "r2r": "0.0.0-beta.0"
```
Comment (ret2libc) [inline package.json]: Why?

Comment (ret2libc) [reply package.json]: And why did you add a package-lock.json file? Please remove both those changes

File: test/new/db/tools/rabin2 [hunk 28]
```diff
@@ -104,12 +104,12 @@ EXPECT=<<EOF
 EOF
 RUN
 
-NAME=rabin2 -z pe
```
Comment (ret2libc) [inline test/new/db/tools/rabin2]: Why did you remove this test?

File: test/new/db/tools/rabin2 [hunk 29]
```diff
@@ -417,8 +417,7 @@ RUN
 NAME=rabin2 -g
 FILE=../bins/elf/analysis/x86-helloworld-gcc
 CMDS=!rabin2 -g ${R2_FILE}
-EXPECT=<<EOF
-[Sections]
+EXPECT="[Sections]
```
Comment (ret2libc) [inline test/new/db/tools/rabin2]: Did you change <<eof in " manually or was it r2r that did it?

File: test/new/db/tools/rabin2 (hunk 28 above)
Comment (kuqadk3) [reply test/new/db/tools/rabin2]: sorry, made some mistakes

Comment (kuqadk3) [reply test/new/db/tools/rabin2]: image
This was automatically generated by r2r but the test keep failing. It's '-g'.

File: test/new/db/tools/rabin2 (hunk 29 above)
Comment (kuqadk3) [reply test/new/db/tools/rabin2]: r2r did that.

File: package.json (hunk 27 above)
Comment (kuqadk3) [reply package.json]: Removed this change

File: test/new/db/tools/rabin2 (hunk 28 above)
Comment (kuqadk3) [reply test/new/db/tools/rabin2]: added it back. I really have no idea why it got deleted. I just run the test and press fix...

File: test/new/db/tools/rabin2 (hunk 29 above)
Comment (radare) [reply test/new/db/tools/rabin2]: uhm we should use <<EOF , the "" syntax is deprecated in r2r.v

Comment (kuqadk3) [reply test/new/db/tools/rabin2]: Changed it back :)

File: test/new/db/tools/rabin2 [hunk 30]
```diff
@@ -625,7 +633,19 @@ Version need section '.gnu.version_r' contains 1 entries:
...
+PHDR         
+INTERP       .interp 
+LOAD0        .interp .note.ABI_tag .note.gnu.build_id .gnu.hash .dynsym .dynstr .gnu.version .gnu.version_r .rel.dyn .rel.plt .init .plt .text .fini .rodata .eh_frame_hdr .eh_frame 
+LOAD1        .init_array .fini_array .jcr .dynamic .got .got.plt .data .bss 
+DYNAMIC      .dynamic 
+NOTE         .note.ABI_tag .note.gnu.build_id 
+GNU_EH_FRAME .eh_frame_hdr 
+GNU_STACK    
+ehdr         
+"
```
[the remark this answers is not in the document]
Comment (kuqadk3) [reply test/new/db/tools/rabin2]: Yes, then I press "fix", then I have this changes.

**Comment by ret2libc** [comment]
If you see Travis it is still red. There are other tests to fix please.

**Comment by kuqadk3** [comment]
> If you see Travis it is still red. There are other tests to fix please.

I don't know how to fix this... https://travis-ci.com/radareorg/radare2/jobs/292257924
image

**Comment by kuqadk3** [comment]
Maybe it was because of this one?
I tried to put both spaces and tabs into the test but it still throws error like this
image

**Comment by ret2libc** [comment]
So, I suggest you:
1) rebase on top of updated master (or merge master into your branch)
2) look at https://travis-ci.com/radareorg/radare2/jobs/292257927#L2567, this is a test you should fix.

If you have problems, tell me ;)

**Comment by kuqadk3** [comment]
> So, I suggest you:
> 
> 1. rebase on top of updated master (or merge master into your branch)
> 2. look at https://travis-ci.com/radareorg/radare2/jobs/292257927#L2567, this is a test you should fix.
> 
> If you have problems, tell me ;)



> Maybe it was because of this one?
> I tried to put both spaces and tabs into the test but it still throws error like this
> image

Since it's highlighting some spaces, I tried to put spaces into those places but it still throws the same error.

File: test/new/db/tools/rabin2 [hunk 31]
```diff
@@ -625,7 +633,19 @@ Version need section '.gnu.version_r' contains 1 entries:
...
+PHDR         
+INTERP       .interp 
+LOAD0        .interp .note.ABI_tag .note.gnu.build_id .gnu.hash .dynsym .dynstr .gnu.version .gnu.version_r .rel.dyn .rel.plt .init .plt .text .fini .rodata .eh_frame_hdr .eh_frame 
+LOAD1        .init_array .fini_array .jcr .dynamic .got .got.plt .data .bss 
+DYNAMIC      .dynamic 
+NOTE         .note.ABI_tag .note.gnu.build_id 
+GNU_EH_FRAME .eh_frame_hdr 
+GNU_STACK    
+ehdr         
+<<EOF
```
Comment (ret2libc) [inline test/new/db/tools/rabin2]: this should be `EOF` not `<<EOF`. If you look at all the other tests you see they use `EOF`

File: test/new/db/tools/rabin2 [hunk 32]
```diff
@@ -625,7 +633,19 @@ Version need section '.gnu.version_r' contains 1 entries:
...
+--------------------
+PHDR         
+INTERP       .interp 
+LOAD0        .interp .note.ABI_tag .note.gnu.build_id .gnu.hash .dynsym .dynstr .gnu.version .gnu.version_r .rel.dyn .rel.plt .init .plt .text .fini .rodata .eh_frame_hdr .eh_frame 
+LOAD1        .init_array .fini_array .jcr .dynamic .got .got.plt .data .bss 
+DYNAMIC      .dynamic 
+NOTE         .note.ABI_tag .note.gnu.build_id 
+GNU_EH_FRAME .eh_frame_hdr 
+GNU_STACK    
+ehdr
```
Comment (ret2libc) [inline test/new/db/tools/rabin2]: add a newline after this line

**Comment by kuqadk3** [comment]
Finally saw green. Thanks @ret2libc and @radare a lot!

**Comment by ret2libc** [comment]
Thanks a lot for all the changes you did! Next PRs will be easier for sure ;)

**Comment by ret2libc** [comment]
@radare please approve and merge

File: test/new/db/tools/rabin2 [hunk 33]
```diff
@@ -104,6 +104,14 @@ EXPECT=<<EOF
 EOF
 RUN
 
+NAME=rabin2 -SSS
+FILE=../bins/elf/analysis/x86-helloworld-gcc
+CMDS=<<EOF
+!rabin2 -SSS ${R2_FILE}|wc -l|awk "{print \$1}"
+EOF
+EXPECT='13'
```
Comment (kazarmy) [inline test/new/db/tools/rabin2]:
```suggestion
EXPECT=13
```

The `..` syntax is deprecated and might not work in the future.

**Comment by kazarmy** [comment]
> I don't know how to fix this... https://travis-ci.com/radareorg/radare2/jobs/292257924
> image

This is probably caused by https://transfer.sh/ being down. No need to worry about this but please check whether there are any other errors.

**Comment by kazarmy** [comment]
And yes the autofix feature is useful but not quite 100% correct.

File: test/new/db/tools/rabin2 (hunk 33 above)
Comment (kuqadk3) [reply test/new/db/tools/rabin2]: Any suggestion on how to fix '..' syntax?
Besides that, I saw other test cases used it as well

**Comment by kuqadk3** [comment]
> > I don't know how to fix this... https://travis-ci.com/radareorg/radare2/jobs/292257924
> > image
> 
> This is probably caused by https://transfer.sh/ being down at the time. No need to worry about this but please check whether there are any other errors.

That's why I posted it at first, to show that it's because of transfer.sh so that build fails. But after that, it was still failed because of other tests that were fixed.
